### DISCUSSION TITLE
Validate variables from other source

### DISCUSSION THREAD

**Discussion by ZCH2005** [eclipse-langium/langium#1491] [Q&A] [answered]
We are exploring Langium  to see if it can be used in our IDE. Our language grammar is similar to Lox: https://github.com/TypeFox/langium-lox/blob/main/src/language-server/lox.langium. 
 
But in our IDE, the variable declarations are not allowed in editor. The variables infomation (name, type...) can be obtained as different format, like JSON. How to let Langium know those variables? Is it enough to only implement our own scope provider? Shall we also need to implement our own  NameProvider and ScopeComputation?

**Comment by msujew** [comment]
Hey @ZCH2005,

have you seen https://github.com/eclipse-langium/langium/discussions/1329? I think the question is fairly similar. The anwser I posted in https://github.com/eclipse-langium/langium/discussions/1329#discussioncomment-8023327 should help you along.

**Comment by ZCH2005** [comment]
Hi @msujew,

Thank  your very much for the reply.  
After reading your comments on [#1329](https://github.com/eclipse-langium/langium/discussions/1329), and then following the tutorial of [buildin-libraries](https://langium.org/guides/builtin-library/), I updated the vs code extension in Lox sample to monitor a file in which Lox variables are defined (e.g., var aaa: number). After launching the vs code extension host, there is no error to use the variables defined in the file. 
   

>    fs.watchFile("C:\\temp\\test.lox", (eventType, filename) => {
>                buildinLox =  fs.readFileSync("C:\\temp\\test.lox", { encoding: 'utf8', flag: 'r' }); 
>                this.documentBuilder.update([URI.parse('builtin:///library.lox')],[]);
>           }); 

In the last part of the  [buildin-libraries](https://langium.org/guides/builtin-library/), the DslLibraryFileSystemProvider was added to src/extension.ts. But we are using [Monaco Editor Wrapper](https://github.com/TypeFox/monaco-components/blob/main/packages/monaco-editor-wrapper) and I could not figure out how to implement the same DslLibraryFileSystemProvider in Monaco Editor Wrapper, do you have any samples?

**Reply by msujew** [reply, marked as answer]
> After launching the vs code extension host, there is no error to use the variables defined in the file.

So you're saying it works? That's great!

> I could not figure out how to implement the same DslLibraryFileSystemProvider in Monaco Editor Wrapper, do you have any samples?

You are best served to [ask the maintainers directly](https://github.com/TypeFox/monaco-languageclient/discussions).
